proposed/http-handlers/request-handlers-meta.md

```diff
@@ -275,6 +276,10 @@ and a request handler and must return a response. The middleware may:
 - Create and return a response without passing the request to the request handler,
   thereby handling the request itself.
 
+It is expected that a middleware dispatching system will use the request handler
+to delegate response creation to the next available middleware. The last middleware
+would then act as a domain gateway to execute application code.
```

Comment:
This still doesn't help me.  I need to see (probably in the meta doc) a code sample of how I'd put 2-3 middlewares and response handlers together.  Right now, the middlware signature wants a response handler for its 3rd parameter, and so I don't know how I'd string together a series of middlewares akin to what I'd do in Stack PHP et al.  I see a type mismatch and my brain segfaults. :smile:

Reply:
The short answer is: this is really hard to explain simply. This is a job better left to vendor implementations. Composing a middleware stack involves having a handler that points back to the stack, as well as a "fail condition" response.

https://github.com/northwoods/broker/blob/1.0.0/src/RequestHandler.php

Reply:
Since to pass the PSR we need 2 working implementation, maybe we can link those here in the meta?

Reply:
@Crell — regarding this:

>  I need to see (probably in the meta doc) a code sample of how I'd put 2-3 middlewares and response handlers together. 

The problem is that _there's more than one way to do it_. In fact, there are a ton of ways to do it.

The most common way is to create a handler that acts as a queue:

```php
class Queue implements RequestHandlerInterface
{
    private $middleware = [];
    private $fallbackResponse;
    
    public function __construct(ResponseInterface $fallbackResponse)
    {
        $this->fallbackResponse = $fallbackResponse;
    }
    
    public function add(MiddlewareInterface $middleware)
    {
        $this->middleware[] = $middleware;
    }
    
    public function handle(ServerRequestInterface $request) : ResponseInterface
    {
        if (0 === count($this->middleware)) {
            return $this->fallbackResponse;
        }
        
        $middleware = array_shift($this->middleware);
        return $middleware->process($request, $this);
    }
}
```

An application bootstrap might then look like this:

```php
// Create instance
$app = new Queue(new Response());

// add one or more middleware:
$app->add(new CorsMiddleware());
$app->add(new FastRouteMiddleware());

// execute it:
$response = $app->handle(ServerRequestFactory::fromGlobals());
```

Each middleware then is passed the `$app` instance as the request handler, and it then determines if it has more middleware to execute, or if it should return its fallback response.

However, you could also write middleware that accepts other middleware via the constructor, and executes it directly:

```php
class Foo implements MiddlewareInterface
{
    private $app;
    
    public function __construct(MiddlewareInterface $app)
    {
        $this->app = $app;
    }
    
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
    {
        $response = $this->app->process($request, $handler);
        // do something with the response...
        return $response;
    }
}
```

The above scenario is similar to how StackPHP (and its predecessor, Rack) operates. It requires that you create your "application" from the inside out, with the innermost middleware being passed to the next layer up, and so on and so forth. The innermost would then either use the request handler to create a response, or create one itself and return it. In your bootstrap, you then would execute the outermost layer:

```php
$requestHandler = new class implements RequestHandlerInterface
{
    public function handle(ServerRequestInterface $request) : ResponseInterface
    {
        return new Response();
    }
}

$response = $outermost->process(ServerRequestFactory::fromGlobals(), $requestHandler);
```

The above two scenarios are the most common. The first is the direction most are going due to the fact that it makes composition easier; you do not need to compose middleware to which you delegate directly within other middleware, just the dependencies on which it operates.

Are these the sort of examples you are looking for?

Reply:
Yes!  My understanding (from previous drafts, most likely) was that you'd somehow pass the $next middleware directly to each, or you're passing along a stack of them, yet neither seemed to be the case.  I see now where that ended up: Externalizing the list of middlewares out of the spec entirely to the Queue class above.  Which... is going to be calling itself doubly-recursively, so it's not really a queue but still a stack.  The call stack in which every other level is the Queue object is not at all obvious and I'd not have stumbled across that on my own.  (It also makes me a little concerned about stack depth and backtrace depth, but that is perhaps because I am used to Drupal 8's super-deep backtraces.)

However, and this is why I wanted to drill down on this front, doesn't it mean a concrete class implemented for the StackPHP-style wouldn't work with the Queue-ish version?  It's $next is provided by the constructor, so it's not going to use what's passed to it in `process()`.  Are they really mix-and-matchable?

(This is the sort of "implementer guidance" I feel needs to be covered, either in the spec or the metadoc.)

Reply:
> doesn't it mean a concrete class implemented for the StackPHP-style wouldn't work with the Queue-ish version? It's $next is provided by the constructor, so it's not going to use what's passed to it in process(). Are they really mix-and-matchable?

They'll still work. If you pass a stack-style middleware into a queue, then:

- you will need to pass some middleware implementation to its constructor
- any items added to the queue after it... will never be executed UNLESS the stack-style middleware calls on its request handler.
- if the stack-style middleware calls on the middleware implementation it composes, it will be passing it the request handler as well, and everything continues merrily on its way.

If you pass a queue-style middleware to a stack-style workflow, if it calls its request handler, then it will get a response back. Considering such middleware will not accept other middleware to the constructor, this means you would likely be using it as an inner-most layer.

So, interoperability is still available. However, stack-style middleware will have some limitations to how it composes non-stack-style middleware, due to the convention it uses for composing the next middleware. Which is why most middleware runners are going the route of using the request handler as a queue/stack/what-have-you for the purpose of composing layers, vs. the convention-based approach of nesting middleware within one another.

If you want, I can write up the above for the metadoc, if you feel it provides the implementer guidance you feel is missing.